### PR TITLE
feat: Fallback values

### LINKED ISSUE #27: Support for fallback values
> If a browser doesn’t recognize a property or its corresponding value, the browser will ignore the property altogether. When this happens, the browser uses (or falls back) to the previous value it finds.
>
> This is the easiest way to provide a fallback. Here’s an example:
>
> ```css
> .layout {
>   display: block;
>   display: grid;
> }
> ```
>
> In this example, browsers that support CSS Grid will use `display: grid`. Browser doesn’t support CSS Grid will fall back to `display: block`.

https://zellwk.com/blog/older-browsers-css/

---

As we use objects to declare styles we can't simply do:

```ts
// ⚠️ this is not real code, this does not work
makeStyles({
  root: {
    color: "red",
    color: "rgba(0, 0, 0, 0.5)"
  }
});
```

We can use the similar approach as [`fela-plugin-fallback-value`](https://www.npmjs.com/package/fela-plugin-fallback-value) with arrays or something else.

The approach to fallback values should be tested and documented.

### PR DESCRIPTION
**Pull request by miroslavstastny**
Add support for fallback values.
Fixes #27.

Any CSS property accepts an array of values which are all added to the styles.
Every browser will use the latest valid value (which might be a different one in different browsers, based on supported CSS in that browser): 

```js
import { makeStyles } from '@griffel/react';

const useClasses = makeStyles({
  root: {
    overflowY: ['scroll', 'overlay'],
  },
});
```

Produces following CSS:

```css
.f1qdoogn{
  overflow-y: scroll;  /* Fallback for browsers which do not support overflow: overlay */
  overflow-y: overlay; /* Used by browsers which support overflow: overlay */
}
```

### CODE REVIEW

File: packages/core/src/runtime/resolveStyleRules.ts [hunk 1]
```diff
@@ -171,6 +171,67 @@ export function resolveStyleRules(
...
+
+      const rtlDefinitions =
+        (rtlValue && [{ key: property, value: rtlValue }]) || value.map(v => convertProperty(property, v!)); // FIXME
+
+      if (process.env.NODE_ENV !== 'production') {
+        const rtlPropertyConsistent =
+          rtlDefinitions.length === 0 || !rtlDefinitions.some(v => v.key !== rtlDefinitions[0].key);
+
+        if (!rtlPropertyConsistent) {
+          throw new Error(
```
Comment (miroslavstastny) [inline packages/core/src/runtime/resolveStyleRules.ts:194]: we should rather warn and ignore the whole property

File: packages/core/src/shorthands/types.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import * as CSS from 'csstype';
+import { GriffelStylesCSSValue, ValueOrArray } from '../types';
+
+export type BorderWidthInput = ValueOrArray<CSS.Property.BorderWidth<GriffelStylesCSSValue>>;
+export type BorderStyleInput = ValueOrArray<CSS.Property.BorderStyle>;
+export type BorderColorInput = ValueOrArray<CSS.Property.BorderColor>;
+
+export type BorderRadiusInput = ValueOrArray<CSS.Property.BorderRadius<GriffelStylesCSSValue>>;
```
[the remark this answers is not in the document]
Comment (miroslavstastny) [reply packages/core/src/shorthands/types.ts:8]: The same applies to `margin` and `padding`

File: apps/website/docs/react/api/make-styles.md [hunk 3]
```diff
@@ -213,3 +213,29 @@ const useClasses = makeStyles({
...
+
+Any CSS property accepts an array of values which are all added to the styles.
+Every browser will use the latest valid value (which might be a different one in different browsers, based on supported CSS in that browser): 
+
+```js
+import { makeStyles } from '@griffel/react';
+
+const useClasses = makeStyles({
+  root: {
+    overflowY: ['scroll', 'overlay'],
```
Comment (miroslavstastny) [inline apps/website/docs/react/api/make-styles.md:229]: Ironically, csstype typings currently do not accept `overlay` as a valid value for `overflow` 😢

Comment (layershifter) [reply apps/website/docs/react/api/make-styles.md:229]: Yup, https://github.com/frenic/csstype/issues/62

File: apps/website/docs/react/api/make-styles.md [hunk 4]
```diff
@@ -213,3 +213,29 @@ const useClasses = makeStyles({
...
+```
+
+<OutputTitle>Produces following CSS...</OutputTitle>
+
+```css
+.f1qdoogn{
+  overflow-y: scroll;  /* Fallback for browsers which do not support overflow: overlay */
+  overflow-y: overlay; /* Used by browsers which support overflow: overlay */
+}
+```
```
Comment (layershifter) [inline apps/website/docs/react/api/make-styles.md:241]: Would you mind to partially duplicate this to `packages/react/README.md`?

File: packages/core/src/runtime/resolveStyleRules.ts [hunk 5]
```diff
@@ -171,6 +171,65 @@ export function resolveStyleRules(
...
+      const className = hashClassName({
+        media,
+        value: value.map(v => (v ?? '').toString()).join(';'),
+        support,
+        pseudo,
+        property,
+      });
+
+      const rtlDefinitions =
+        (rtlValue && [{ key: property, value: rtlValue }]) || value.map(v => convertProperty(property, v!));
```
[the remark this answers is not in the document]
Comment (miroslavstastny) [reply packages/core/src/runtime/resolveStyleRules.ts:187]: This should never happen.
The only moment `rtlValue` is set is when the `resolveStyleRules` is called recursively for `animationName` property.

In that case the type of value is `string` and `property` is `animationName`. So the code will never get to the fallback values branch.

In this PR, I'd suggest to remove the `rtlValue` check.

TBH, I do not like the `rtlValue` argument, what do you think about refactoring the function in a separate PR?
1. remove the `rtlValue` argument
2. move `if (typeof value === 'string' || typeof value === 'number') {` branch to a separate private function
3. in `animationName` branch, call the function from the previous step

File: packages/core/src/runtime/resolveStyleRules.ts [hunk 6]
```diff
@@ -171,6 +171,65 @@ export function resolveStyleRules(
         cssRulesByBucket,
         rtlAnimationNames.join(', '),
       );
+    } else if (Array.isArray(value) && value.length > 0) {
+      // not animationName property but array in the value => fallback values
+
+      const key = hashPropertyKey(pseudo, media, support, property);
+      const className = hashClassName({
+        media,
+        value: value.map(v => (v ?? '').toString()).join(';'),
```
[the remark this answers is not in the document]
Comment (miroslavstastny) [reply packages/core/src/runtime/resolveStyleRules.ts:180]: At a price of additional `Array.isArray()` on a hot path.

[the remark this answers is not in the document]
Comment (layershifter) [reply packages/core/src/runtime/resolveStyleRules.ts:180]: I see, let's leave it as it is then.

File: packages/core/src/runtime/compileCSS.ts [hunk 7]
```diff
@@ -63,14 +63,18 @@ export function compileCSS(options: CompileCSSOptions): [string /* ltr definitio
...
+    : `{ ${hyphenateProperty(property)}: ${value}; }`;
 
   let rtlClassNameSelector: string | null = null;
   let rtlCSSDeclaration: string | null = null;
 
   if (rtlProperty && rtlClassName) {
     rtlClassNameSelector = `.${rtlClassName}`;
-    rtlCSSDeclaration = `{ ${hyphenateProperty(rtlProperty)}: ${rtlValue}; }`;
+    rtlCSSDeclaration = Array.isArray(rtlValue)
+      ? `{ ${rtlValue?.map(v => `${hyphenateProperty(rtlProperty)}: ${v}`).join(';')}; }`
```
[the remark this answers is not in the document]
Comment (miroslavstastny) [reply packages/core/src/runtime/compileCSS.ts]: Removed

File: change/@griffel-core-6035301a-4045-4891-8c41-5d91d4e0a08e.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
```
[the remark this answers is not in the document]
Comment (miroslavstastny) [reply change/@griffel-core-6035301a-4045-4891-8c41-5d91d4e0a08e.json]: Change files re-generated.

File: packages/core/src/runtime/compileCSS.test.ts [hunk 9]
```diff
@@ -142,11 +173,11 @@ describe('compileCSS', () => {
...
-      Array [
-        "body .foo{padding-left:10px;}",
-        "body .rtl-foo{padding-right:10px;}",
-      ]
-    `);
+              Array [
+                "body .foo{padding-left:10px;}",
+                "body .rtl-foo{padding-right:10px;}",
+              ]
+          `);
```
[the remark this answers is not in the document]
Comment (miroslavstastny) [reply packages/core/src/runtime/compileCSS.test.ts]: Reverted, sorry :/

File: packages/core/src/runtime/resolveStyleRules.ts [hunk 10]
```diff
@@ -171,6 +171,65 @@ export function resolveStyleRules(
         cssRulesByBucket,
         rtlAnimationNames.join(', '),
       );
+    } else if (Array.isArray(value) && value.length > 0) {
+      // not animationName property but array in the value => fallback values
```
[the remark this answers is not in the document]
Comment (miroslavstastny) [reply packages/core/src/runtime/resolveStyleRules.ts]: Warning added, UT updated.

File: packages/core/src/runtime/resolveStyleRules.ts [hunk 11]
```diff
@@ -171,6 +171,65 @@ export function resolveStyleRules(
...
+      if (process.env.NODE_ENV !== 'production') {
+        const rtlPropertyConsistent = !rtlDefinitions.some(v => v.key !== rtlDefinitions[0].key);
+
+        if (!rtlPropertyConsistent) {
+          console.error(
+            'makeStyles(): mixing CSS fallback values which result in multiple CSS properties in RTL is not supported.',
+          );
+          continue;
+        }
+      }
```
[the remark this answers is not in the document]
Comment (miroslavstastny) [reply packages/core/src/runtime/resolveStyleRules.ts]: Correct, fixed.

File: apps/website/docs/react/api/make-styles.md (hunk 4 above)
Comment (miroslavstastny) [reply apps/website/docs/react/api/make-styles.md:241]: README updated

**Review by layershifter — APPROVED** [review]

**Pull request merged**